src/runtime/contrib/tensorrt/tensorrt_builder.cc

```diff
@@ -92,54 +91,109 @@ GetOpConverters() {
   map->emplace("floor", std::make_shared<UnaryOpConverter>());
   map->emplace("strided_slice", std::make_shared<StridedSliceOpConverter>());
+  map->emplace("split", std::make_shared<SplitOpConverter>());
 #else
   map->emplace("clip", std::make_shared<ClipLegacyOpConverter>());
 #endif
 #if TRT_VERSION_GE(6, 0, 1)
   map->emplace("image.resize", std::make_shared<ResizeOpConverter>());
+  map->emplace("nn.upsampling", std::make_shared<UpsamplingOpConverter>());
 #endif
   return map;
 }
 
-TensorRTBuilder::TensorRTBuilder(const std::vector<DLTensor*>& args)
-    : execution_args_(args) {
+TensorRTBuilder::TensorRTBuilder(runtime::TensorRTLogger* logger,
+    const std::vector<DLTensor*>& args, size_t max_workspace_size)
+    : execution_args_(args), max_workspace_size_(max_workspace_size) {
   // Create TRT builder and network.
-  static runtime::TensorRTLogger logger;
-  builder_ = nvinfer1::createInferBuilder(logger);
+  builder_ = nvinfer1::createInferBuilder(*logger);
+#if TRT_VERSION_GE(6, 0, 1)
+  // Use INetworkV2 with explicit batch.
+  const auto flags =
+      1U << static_cast<uint32_t>(
+          nvinfer1::NetworkDefinitionCreationFlag::kEXPLICIT_BATCH);
+  network_ = builder_->createNetworkV2(flags);
+#else
+  // Use INetwork with implicit batch.
   batch_size_ = args[0]->shape[0];
   builder_->setMaxBatchSize(batch_size_);
-  const size_t workspace_size =
-      dmlc::GetEnv("TVM_TENSORRT_MAX_WORKSPACE_SIZE", size_t(1) << 31);
-  builder_->setMaxWorkspaceSize(workspace_size);
+  builder_->setMaxWorkspaceSize(max_workspace_size_);
   const bool use_fp16 = dmlc::GetEnv("TVM_TENSORRT_USE_FP16", false);
   builder_->setFp16Mode(use_fp16);
   network_ = builder_->createNetwork();
+#endif
 }
 
-runtime::TrtEngineAndContext TensorRTBuilder::BuildEngine(const Expr& expr) {
-  // Process graph and create INetworkDefinition.
-  VisitExpr(expr);
+void TensorRTBuilder::ProcessInputs(const Function& func) {
+  // All input names in order. This order matches that of execution args.
+  for (size_t i = 0; i < func->params.size(); i++) {
+    network_input_names_.push_back(func->params[i]->name_hint());
+    network_input_map_[func->params[i]->name_hint()] = i;
+  }
+  // Assume all inputs are real to start. If an input is baked into the TRT
+  // engine, we will set the entry in this array to true.
+  network_input_is_baked_.assign(func->params.size(), false);
+}
+
+void TensorRTBuilder::ProcessOutputs(const Expr& expr) {
   // Mark outputs.
   auto it = node_output_map_.find(expr.operator->());
   CHECK(it != node_output_map_.end()) << "Output was not found.";
   auto network_outputs = it->second;
-  std::vector<std::string> network_output_names;
   for (size_t i = 0; i < network_outputs.size(); ++i) {
     CHECK(network_outputs[i].type == kTensor);
     auto out_tensor = network_outputs[i].tensor;
+
+    // Workaround to support duplicate outputs
+    if (out_tensor->isNetworkOutput()) {
```

Comment:
Since the duplicated output issue has been resolved, we might want to put a TODO here so that we will remember to remove this workaround after rebasing.

Reply:
Thanks for the review! This code is already on the branch with the fix, so the workaround should no longer be needed. However, maybe its a good idea to keep it for now since it doesn't cause any harm and may prevent some crashes if duplicate outputs occured for some other reason we don't know about. I will update the comment though.